Segfault on --help with --tlsverify if no ca certificate is present

BUG REPORT INFORMATION
---------------------------------------------------

**Description**

Docker segfaults when trying to list view the helptext (`--help`) when using tls verification without a CA certificate.

**Steps to reproduce the issue:**
1. Make sure there is no ca.pem in `$HOME/.docker/`
2. Execute `docker --tlsverify --help`

**Describe the results you received:**

```docker --tlsverify --help
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x8d8605]

goroutine 1 [running]:
github.com/docker/cli/vendor/github.com/docker/docker/client.(*Client).ClientVersion(0x0, 0x2507800, 0x0)
	/go/src/github.com/docker/cli/vendor/github.com/docker/docker/client/client.go:321 +0x5
main.areSubcommandsSupported(0xc42058a240, 0x24dc060, 0xc420101b80, 0x24ca720, 0xc4204b9440)
	/go/src/github.com/docker/cli/cmd/docker/docker.go:309 +0x60
main.isSupported(0xc42058a240, 0x24dc060, 0xc420101b80, 0x7ffe6135d7d0, 0x24ba280)
	/go/src/github.com/docker/cli/cmd/docker/docker.go:260 +0x43
main.setHelpFunc.func1(0xc42058a240, 0xc4200100a0, 0x2, 0x2)
	/go/src/github.com/docker/cli/cmd/docker/docker.go:90 +0x72
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc42058a240, 0xc420455470, 0xc420455410, 0xc42000e200)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:790 +0x5bb
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).Execute(0xc42058a240, 0xc42058a240, 0x24cef20)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:738 +0x2b
main.main()
	/go/src/github.com/docker/cli/cmd/docker/docker.go:162 +0xd0
```

**Describe the results you expected:**
The same warning you get when executing for example `docker --tlsverify info` without a ca.pem in `$HOME/.docker/`: `could not read CA certificate "$HOME/.docker/ca.pem": open $HOME/.docker/ca.pem: no such file or directory`

**Additional information you deem important (e.g. issue happens only occasionally):**
Tested on both Arch Linux and Debian Stretch

**Output of `docker version`:**
Arch Linux:

```
Client:
 Version:      18.05.0-ce
 API version:  1.37
 Go version:   go1.10.2
 Git commit:   f150324782
 Built:        Wed May 16 22:27:45 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.05.0-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.10.2
  Git commit:   f150324782
  Built:        Wed May 16 22:28:17 2018
  OS/Arch:      linux/amd64
  Experimental: false

```

Debian Stretch:

```
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:17:14 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:15:24 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Output of `docker info`:**

Arch Linux:

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 33
Server Version: 18.05.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.16.13-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.698GiB
Name: <left out>
ID: DHWN:5OLZ:O6NU:FBTD:3XTA:PI2N:E3ON:N4NT:AI2Z:BSSV:JJS3:ETLW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

Debian Stretch:

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-6-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 996.4MiB
Name: <left out>
ID: ABAF:EW3C:EAZP:LFMA:2KRF:4L2P:42U2:6TWF:ZN2D:NVQZ:SHVA:UH75
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Arch Linux on a physical device
Debian Stretch on a VPS

--- a/cmd/docker/docker.go
+++ b/cmd/docker/docker.go
@@ -83,7 +83,9 @@ func setFlagErrorFunc(dockerCli *command.DockerCli, cmd *cobra.Command, flags *p
 	// is called.
 	flagErrorFunc := cmd.FlagErrorFunc()
 	cmd.SetFlagErrorFunc(func(cmd *cobra.Command, err error) error {
-		initializeDockerCli(dockerCli, flags, opts)
+		if err := initializeDockerCli(dockerCli, flags, opts); err != nil {
+			return err
+		}
 		if err := isSupported(cmd, dockerCli); err != nil {
 			return err
 		}
@@ -94,7 +96,10 @@ func setFlagErrorFunc(dockerCli *command.DockerCli, cmd *cobra.Command, flags *p
 func setHelpFunc(dockerCli *command.DockerCli, cmd *cobra.Command, flags *pflag.FlagSet, opts *cliflags.ClientOptions) {
 	defaultHelpFunc := cmd.HelpFunc()
 	cmd.SetHelpFunc(func(ccmd *cobra.Command, args []string) {
-		initializeDockerCli(dockerCli, flags, opts)
+		if err := initializeDockerCli(dockerCli, flags, opts); err != nil {
+			ccmd.Println(err)
+			return
+		}
 		if err := isSupported(ccmd, dockerCli); err != nil {
 			ccmd.Println(err)
 			return
@@ -123,7 +128,9 @@ func setValidateArgs(dockerCli *command.DockerCli, cmd *cobra.Command, flags *pf
 
 		cmdArgs := ccmd.Args
 		ccmd.Args = func(cmd *cobra.Command, args []string) error {
-			initializeDockerCli(dockerCli, flags, opts)
+			if err := initializeDockerCli(dockerCli, flags, opts); err != nil {
+				return err
+			}
 			if err := isSupported(cmd, dockerCli); err != nil {
 				return err
 			}
@@ -132,13 +139,15 @@ func setValidateArgs(dockerCli *command.DockerCli, cmd *cobra.Command, flags *pf
 	})
 }
 
-func initializeDockerCli(dockerCli *command.DockerCli, flags *pflag.FlagSet, opts *cliflags.ClientOptions) {
-	if dockerCli.Client() == nil { // when using --help, PersistentPreRun is not called, so initialization is needed.
-		// flags must be the top-level command flags, not cmd.Flags()
-		opts.Common.SetDefaultOptions(flags)
-		dockerPreRun(opts)
-		dockerCli.Initialize(opts)
+func initializeDockerCli(dockerCli *command.DockerCli, flags *pflag.FlagSet, opts *cliflags.ClientOptions) error {
+	if dockerCli.Client() != nil {
+		return nil
 	}
+	// when using --help, PersistentPreRun is not called, so initialization is needed.
+	// flags must be the top-level command flags, not cmd.Flags()
+	opts.Common.SetDefaultOptions(flags)
+	dockerPreRun(opts)
+	return dockerCli.Initialize(opts)
 }
 
 // visitAll will traverse all commands from the root.

